File: Leetcode/Leetcode-DSA-course/Linked-List/24-swap-nodes-in-pairs/swap-nodes-in-pairs.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* swapPairs(ListNode* head) {
        if (head == nullptr || head->next == nullptr)
            return head;
        ListNode* prev = nullptr;
        ListNode* dummy = head->next;
        while (head != nullptr && head->next != nullptr) {
            if (prev != nullptr)
                prev->next = head->next;
            prev = head;

            ListNode* nextNode = head->next->next;
            head->next->next = head;
            head->next = nextNode;
            head = nextNode;
        }
        return dummy;
    }
};

File: Leetcode/Leetcode-DSA-course/Linked-List/24-swap-nodes-in-pairs/swap-nodes-in-pairs.ts
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */

function swapPairs(head: ListNode | null): ListNode | null {
    if (!head || !head.next)
        return head;
    let prev = null;
    let dummy = head.next;
    while (head && head.next) {
        if (prev) prev.next = head.next;
        prev = head;
        let nextNode = head.next.next;
        head.next.next = head;
        head.next = nextNode;
        head = nextNode;
    }
    return dummy;
};